### PR TITLE
Simplify IsNotNull and IsNull expression

### PR DESCRIPTION
**Pull request by byteink**
# Which issue does this PR close?

None

# Rationale for this change

`expr_non_null IS NOT NULL`  is always true
`expr_non_null IS NULL`  is always false

They are a type of predicate that is either unnecessary or impossible.
By simplifying them, we can gain more opportunities for optimization.  

For example：
1. `SELECT * FROM t WHERE col_non_null IS NOT NULL` --(Simplify)--> `SELECT * FROM t WHERE TRUE` --(EliminateFilter)--> `SELECT * FROM t`
  
2.  `SELECT * FROM t WHERE col_non_null IS NULL` --(Simplify)--> `SELECT * FROM t WHERE FALSE` --(EliminateFilter)--> `<EmptyRelation>`

# Are these changes tested?
Yes

# Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs [hunk 1]
```diff
@@ -1161,6 +1161,12 @@ impl<'a, S: SimplifyInfo> TreeNodeRewriter for Simplifier<'a, S> {
                 lit(!negated)
             }
 
+            // a IS NOT NULL --> true, if a is not nullable
+            Expr::IsNotNull(expr) if !info.nullable(&expr)? => lit(true),
```
Comment (comphead) [inline datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:1165]: Thanks @byteink. would you mind adding more description? Why the simplification needed and what exactly problem it solves? is it a bug or something?

Comment (jackwener) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:1165]:
```sql
create table test(a int not null);

select a is null
```

we can convert it into 

```sql
select false
```

Comment (byteink) [reply datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs:1165]: @comphead It is not a bug. I've added some descriptions.

**Review by alamb — APPROVED** [review]
This looks great -- thank you @byteink  and @comphead  and @jackwener

**Pull request merged**